### PR TITLE
new module that allows for address service object creation

### PR DESCRIPTION
**Pull request by ivanbojer**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
panos_address

##### ANSIBLE VERSION

```
ansible 2.1.0.0 
config file = 
configured module search path = Default w/o overrides
```

##### SUMMARY
New panos module that allows for creation of service address service object of different types [IP Range, FQDN, or IP Netmask].

### CODE REVIEW

**Comment by ivanbojer** [comment]
ready_for_review

File: lib/ansible/modules/network/panos/panos_address.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+description:
+    - Create address service object of different types [IP Range, FQDN, or IP Netmask].
+author: "Luigi Mori (@jtschichold), Ken Celenza (@itdependsnetworks), Ivan Bojer (@ivanbojer)"
+version_added: "2.3"
+requirements:
+    - pan-python
+options:
+    ip_address:
+        description:
+            - IP address (or hostname) of PAN-OS device
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_address.py]: All descriptions must be written in full sentences, i.e. with capital letters and fullstops.
http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#documentation-block

This applies to multiple options

File: lib/ansible/modules/network/panos/panos_address.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        required: true
+        default: None
+    address_name:
+        description:
+            - name of the address
+        required: true
+        default: None
+    type:
+        description:
+            - ip-netmask, fqdn, ip-range
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_address.py]: The `description:` need to describe the *what*
`choices:` (which needs adding) lists the valid options.

See http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#documentation-block for more info.

File: lib/ansible/modules/network/panos/panos_address.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+            - IP address (or hostname) of PAN-OS device
+        required: true
+    password:
+        description:
+            - password for authentication
+        required: true
+    username:
+        description:
+            - username for authentication
+        required: false
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_address.py]: `required: false` lines can be removed

File: lib/ansible/modules/network/panos/panos_address.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        default: None
+    commit:
+        description:
+            - commit if changed
+        required: false
+        default: true
+'''
+
+EXAMPLES = '''
+# Creates service for port 22
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_address.py]: This comment doesn't seem to be related to the examples

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/panos/panos_address.py (hunk 1 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_address.py]: OK

File: lib/ansible/modules/network/panos/panos_address.py (hunk 3 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_address.py]: OK

File: lib/ansible/modules/network/panos/panos_address.py (hunk 2 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_address.py]: OK

File: lib/ansible/modules/network/panos/panos_address.py (hunk 4 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_address.py]: fixed

**Comment by ivanbojer** [comment]
@gundalow Thank you for the review, I've incorporated all the changes requested.

**Comment by ivanbojer** [comment]
ready_for_review

**Comment by ivanbojer** [comment]
ready_for_review

File: lib/ansible/modules/network/panos/panos_address.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+                              type,
+                              tag)
+    except PanXapiError:
+        exc = get_exception()
+        module.fail_json(msg=exc.message)
+
+    if changed and commit:
+        xapi.commit(cmd="<commit></commit>", sync=True, interval=1)
+
+    module.exit_json(changed=changed, msg="okey dokey")
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_address.py:211]: Is there a better `msg`?

File: lib/ansible/modules/network/panos/panos_address.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'version': '1.0'}
+
+DOCUMENTATION = '''
+---
+module: panos_address
+short_description: create address service object
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_address.py]: Create address service object on PanOS devices 

This may need updating on existing modules

File: lib/ansible/modules/network/panos/panos_address.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        description:
+            - Description of the address object.
+        default: None
+    tag:
+        description:
+            - Tag of the address object.
+        default: None
+    commit:
+        description:
+            - Commit if changed
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_address.py]: Same as on other PR 
Missing full stop, 
Is a commit message added?

File: lib/ansible/modules/network/panos/panos_address.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+requirements:
+    - pan-python
+options:
+    ip_address:
+        description:
+            - IP address (or hostname) of PAN-OS device being configured.
+        required: true
+    username:
+        description:
+            - Username credentials to use for auth.
```
Comment (gundalow) [inline lib/ansible/modules/network/panos/panos_address.py]: Same as other PR

**Review by gundalow — CHANGES_REQUESTED** [review]
The comments I left on your other panos PR also apply here.

File: lib/ansible/modules/network/panos/panos_address.py (hunk 6 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_address.py]: OK

File: lib/ansible/modules/network/panos/panos_address.py (hunk 7 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_address.py]: Ok added full stop.

File: lib/ansible/modules/network/panos/panos_address.py (hunk 5 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_address.py:211]: Not really, the call does not return anything. We are using this message for past 2 years and it is a bit of 'developer justice.' Customers seem to have no issue with it.

**Comment by ivanbojer** [comment]
ready_for_review

File: lib/ansible/modules/network/panos/panos_address.py (hunk 5 above)
Comment (sivel) [reply lib/ansible/modules/network/panos/panos_address.py:211]: I really recommend something better here.  I'm not sure we are in the business of "developer justice".  It just looks childish.  If anything, just remove the msg

Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_address.py:211]: @sivel Our customers use our modules for 2+ years and this is the message they see when things go right. Not one of them complained about it. While you find it childish other find it geeky and cute. I don't plan to remove it from this or any other module.

File: lib/ansible/modules/network/panos/panos_address.py (hunk 8 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_address.py]: OK

**Review by gundalow — APPROVED** [review]
